b2luigi/batch/processes/gbasf2.py

```diff
@@ -0,0 +1,678 @@
+        all jobs are done, return ``JobStatus.successful`` to show that the
+        gbasf2 project succeeded.
+
+        The status of each individual job can be one of::
+
+            [Submitting, Submitted, Received, Checking, Staging, Waiting, Matched, Rescheduled,
+             Running, Stalled, Completing, Done, Completed, Failed, Deleted, Killed]
+
+        (Taken from  https://github.com/DIRACGrid/DIRAC/blob/rel-v7r1/WorkloadManagementSystem/Client/JobStatus.py)
+
+        """
+        # If project is does not exist on grid yet, so can't query for gbasf2 project status
+        if not self._check_project_exists():
+            raise RuntimeError(
+                f"Could not find any jobs for project {self.gbasf2_project_name} on the grid.\n" +
+                "Probably there was an error during the project submission when running the gbasf2 command.\n" +
+                "Try if you can run the gbasf2 command used manually in a terminal with gbasf2 set up:\n" +
+                " ".join(self._build_gbasf2_submit_command())
+            )
+
+        job_status_dict = self._get_job_status_dict()
+        n_jobs_by_status = Counter()
+        for _, job_info in job_status_dict.items():
+            n_jobs_by_status[job_info["Status"]] += 1
+
+        # print summary of jobs in project if setting is set and job status changed
+        if (get_setting("gbasf2_print_status_updates", default=True, task=self.task) and
+                n_jobs_by_status != self._n_jobs_by_status):
+            time_string = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+            job_status_string = str(dict(sorted(n_jobs_by_status.items()))).strip("{}")
+            print(f"Jobs in gbasf2 project \"{self.gbasf2_project_name}\" at {time_string}: {job_status_string}")
+        self._n_jobs_by_status = n_jobs_by_status
+
+        n_jobs = len(job_status_dict)
+        n_done = n_jobs_by_status["Done"]
+        # note: "Killed" jobs also get moved to "Failed" after a while
+        n_failed = n_jobs_by_status["Failed"]
+        n_in_final_state = n_done + n_failed
+
+        # The gbasf2 project is considered as failed if any of the jobs in it failed.
+        # However, we first try to reschedule thos jobs and only declare it as failed if the maximum number of retries
+        # for reschedulinhas been reached
+        if n_failed > 0:
+            self._on_failure_action()
+            if self.max_retries > 0 and self._reschedule_failed_jobs():
+                return JobStatus.running
+            return JobStatus.aborted
+
+        if n_in_final_state < n_jobs:
+            return JobStatus.running
+
+        # Require all jobs to be done for project success, any job failure results in a failed project
+        if n_done == n_jobs:
+            # download dataset only the first time that we return JobStatus.successful
+            if not self._project_had_been_successful:
+                self._on_first_success_action()
```

Comment:
So the way I solved inititating the project dataset download from the grid after all jobs had been successful is to add the code for that as a side effect into the `get_job_status` method, which on success triggers the `on_first_success_action`. I don't really like that the `get_job_status` does more than just getting the job status. maybe there's a better way to do that, if you know some luigi functions I could implement I would be glad to do it there.

Reply:
Would it work with the luigi event handlers (https://luigi.readthedocs.io/en/stable/tasks.html#events)?
Although I am not sure we call them correctly in every circumstance (and if they are called on the submission side).
Just came to my mind without me thinking about it more clearly. Don't know if you had a look into it already.

Reply:
Didn't know about the event handlers I will look into that, thanks. I only knew about the `on_success` etc. methods of tasks and I didn't know how to hook into that.

Reply:
Just had a brief look on the event handlers. Sorry for procrastinating on that, worked on some other things inbetween. So as far as I understand they operate on class level, that is that you add event handler to `Task` or subclasses of it. They are stored in the static property `_event_callbacks`. In the batch process, I have access to the task instance the event handler is operating on, but I am not sure yet how to add it only this instance (`self.task`).

I could add a generic event handler to all tasks and then put an if-condition into it which checks if there gbasf2 batch is used, e.g.

```python
@luigi.Task.event_handler(luigi.Event.SUCCESS)
def download_dataset_if_batch_is_gbasf2(task):
    if get_setting("batch", task=task) == "gbasf2":
        run_with_gbasf2(["gb2_ds_get", project_name])
```
I don't really find this solution elegant, but maybe better than side-effects in the `get_job_status`. Do you have a better idea, what do you think?

Reply:
Ah ok. Well no, that seems not like a very elegant solution, as you have said it.

Then lets stick with your solution! We can always come back later and refine if we find a better solution.